### DISCUSSION TITLE
Is it possible to define constants?

### DISCUSSION THREAD

**Discussion by martinvonwittich** [bpftrace/bpftrace#2338] [Q&A] [answered]
I've been playing around with a script to display STDOUT/STDERR writes of arbitrary length of a process identified by its `comm`, and so far I've got this working:

```
#!/usr/local/sbin/bpftrace

BEGIN
{
  @debug = 0;
}

uprobe:libc:write /strncmp("net", comm, 3) == 0 && (arg0 == 1 || arg0 == 2)/
{
  /*
    bpftrace defaults to BPFTRACE_STRLEN=64 which is 63 bytes + NUL terminator.
    str() can't retrieve longer strings than BPFTRACE_STRLEN, so we need to 
    stay below this limit.
  */
  $bufsize = 63;
  $maxlen = 2048;
  $i = 0;

  while ($i <= $maxlen && (uint64)$i <= arg2)
  {
    $offset = arg1 + (uint64)$i;
    $len = arg2 - (uint64)$i > (uint64)$bufsize? $bufsize: arg2 - (uint64)$i;
    if (@debug)
    {
      printf("(%s@%d fd %d totallen %d remlen %d offset %d)%s", comm, tid, arg0, arg2, $len, $i, str(arg1 + (uint64)$i, $len))
    }
    else
    {
      printf("%s", str(arg1 + (uint64)$i, $len))
    }
    $i = $i + $bufsize;
  }
}
```

I would like to replace the `arg0 == 1 || arg0 == 2` with STDOUT and STDERR constants, and also `@debug`, `$bufsize` and `$maxlen` (am I assuming correctly that using variables for this purpose wastes valuable stack space?), but I haven't found a way to define constants in the reference. I've also tried `#define` because the reference mentions `#include`, but that also didn't work. Am I missing something, or is this currently impossible?

**Comment by viktormalik** [comment, marked as answer]
Unforunately, `#define` directives are not handled well at the moment. What you could do is to use [positional parameters](https://github.com/iovisor/bpftrace/blob/master/docs/reference_guide.md#9-1--n--positional-parameters) but that probably doesn't exactly suit your needs. Another option would be to preprocess yout `.bt` file with a C preprocessor, which would allow you to use `#define`.

**Reply by martinvonwittich** [reply]
Should I file a feature request for this?

**Reply by viktormalik** [reply]
Yes, that will be useful, thanks.

**Reply by gsingh93** [reply]
> Unforunately, #define directives are not handled well at the moment.

Could you clarify exactly when we should expect them to work? I see there is a very basic test with them: https://github.com/iovisor/bpftrace/blob/master/tests/runtime/basic#L160

**Comment by viktormalik** [comment]
I've had another look into this and using `#define` works for me:
```
#!/usr/local/sbin/bpftrace

#define STDOUT 1
#define STDERR 2
#define DEBUG 0
#define BUFSIZE 63
#define MAXLEN 2048

uprobe:libc:write /strncmp("net", comm, 3) == 0 && (arg0 == STDOUT || arg0 == STDERR)/
{
  /*
    bpftrace defaults to BPFTRACE_STRLEN=64 which is 63 bytes + NUL terminator.
    str() can't retrieve longer strings than BPFTRACE_STRLEN, so we need to 
    stay below this limit.
  */
  $i = 0;

  while ($i <= MAXLEN && (uint64)$i <= arg2)
  {
    $offset = arg1 + (uint64)$i;
    $len = arg2 - (uint64)$i > (uint64)BUFSIZE? BUFSIZE: arg2 - (uint64)$i;
    if (DEBUG)
    {
      printf("(%s@%d fd %d totallen %d remlen %d offset %d)%s", comm, tid, arg0, arg2, $len, $i, str(arg1 + (uint64)$i, $len))
    }
    else
    {
      printf("%s", str(arg1 + (uint64)$i, $len))
    }
    $i = $i + BUFSIZE;
  }
}
```

**Comment by martinvonwittich** [comment]
> I've had another look into this and using `#define` works for me:

Can confirm, `#define` is now working fine for me too!
